lib/experimental/sync/class-gutenberg-http-signaling-server.php

```diff
@@ -0,0 +1,343 @@
+			die( 'no identifier' );
+		}
+		static::$subscriber_id = $_REQUEST['subscriber_id'];
+
+		if ( 'GET' === $_SERVER['REQUEST_METHOD'] ) {
+			static::handle_message_read_request();
+		} else {
+			if ( empty( $_POST ) || empty( $_POST['message'] ) ) {
+				die( 'no message' );
+			}
+			$message = json_decode( wp_unslash( $_POST['message'] ), true );
+			if ( ! $message ) {
+				die( 'no message' );
+			}
+			static::handle_message_operation( $message );
+		}
+
+		static::clean_up_old_connections();
+		exit;
+	}
+
+	/**
+	 * Initializes the paths of the temporary files used.
+	 */
+	private static function initialize_paths() {
+		static::$subscriber_to_messages_path = get_temp_dir() . DIRECTORY_SEPARATOR . 'subscribers_to_messages.txt';
+		// Example: array( 2323232121 => array( 'message hello','handshake message' ) ).
+
+		static::$topics_to_subscribers_path = get_temp_dir() . DIRECTORY_SEPARATOR . 'topics_to_subscribers.txt';
+		// Example: array( 'doc1: array( 2323232121 ), 'doc2: array( 2323232123, 2323232121 ) ).
+
+		static::$subscribers_to_last_connection_path = get_temp_dir() . DIRECTORY_SEPARATOR . 'subscribers_to_last_connection.txt';
+		// Example: array( 2323232121 => 34343433323(timestamp) ).
+	}
+
+	/**
+	 * Reads the contents of $fd and returns them unserialized.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param resource $fd  A file descriptor.
+	 *
+	 * @return array Unserialized contents of fd.
+	 */
+	private static function get_contents_from_file_descriptor( $fd ) {
+		$contents_raw = stream_get_contents( $fd );
+		$result       = array();
+		if ( $contents_raw ) {
+			$result = unserialize( $contents_raw );
+		}
+		return $result;
+	}
+
+
+	/**
+	 * Makes the file descriptor content of $fd equal to the serialization of content.
+	 * Overwrites what was previously in $fd.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param resource $fd      A file descriptor.
+	 * @param array    $content Content to be serialized and written in a file descriptor.
+	 */
+	private static function save_contents_to_file_descriptor( $fd, $content ) {
+		rewind( $fd );
+		$data = serialize( $content );
+		fwrite( $fd, $data );
+		ftruncate( $fd, strlen( $data ) );
+	}
+
+	/**
+	 * Handles a wp_ajax signaling server request of client that wants to retrieve its messages.
+	 *
+	 * It returns the client a response following the
+	 * {@link https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format Event stream format}.
+	 *
+	 * ```
+	 * id: <Event ID>
+	 * retry: <Reconnection time, in ms>
+	 * event: <The type of event>
+	 * data: <The message to be sent>
+	 * ```
+	 */
+	private static function handle_message_read_request() {
+		header( 'Content-Type: text/event-stream' );
+		header( 'Cache-Control: no-cache' );
+		echo 'retry: 3000' . PHP_EOL;
+		$fd = fopen( static::$subscriber_to_messages_path, 'c+' );
+		if( ! $fd ) {
+			die( 'Could not open required file.' );
```

Comment:
In terms of error-handling, the spec doesn't seem to provide ways for the Event stream to encode errors that happened in the server. It sounds like it's up to the protocol to define how to handle this.

What alternatives do we have to `die` when we cannot read a file?

I'm thinking that a simple one could be the following: we always respond to a request, but if we cannot read a file, we double the retry time. This way, the retry time for the client will grow exponentially (we don't want clients to DDoS a server that is having issues) while still gracefully accounting for issues and don't let clients hanging up on requests.

Pseudo-code:

```php
header( 'Content-Type: text/event-stream' );
header( 'Cache-Control: no-cache' );
if( $fd_could_be_opened ) {
  // Normal operation.
  $current_retry = 3000; // reset the retry
  echo 'retry: ' . $current_retry;

  // Add data, event, ...
} else {
   $current_retry = $current_retry * 2;
   echo 'retry: ' + $current_retry;
}
```

Reply:
I added a retry mechanism as suggested but to store the current retry across different requests it was a little complex. I used a cookie.